File: laby/src/main/java/agh/ics/oop/model/Animal.java
package agh.ics.oop.model;



public class Animal implements WorldElement {
    private Vector2d position;
    private MapDirection direction;

    public Animal(){
        this.position = new Vector2d(2,2);
        this.direction = MapDirection.NORTH;
    }

    public  Animal(Vector2d initialPosition){
        this.position = initialPosition;
        this.direction = MapDirection.NORTH;
    }

    @Override
    public String toString() {

        return switch (this.direction)
        {
            case NORTH -> "^";
            case EAST -> ">";
            case SOUTH -> "v";
            case WEST -> "<";
        };


    }
    public boolean isAt(Vector2d position){
        return this.position.equals(position);
    }
    public void move(MoveDirection direction, MoveValidator validator)
    {
        switch (direction)
        {
            case RIGHT -> this.direction = this.direction.next();
            case LEFT -> this.direction = this.direction.previous();
            case FORWARD -> {
                Vector2d newPosition = this.position.add(this.direction.toUnitVector());
                if (validator.canMoveTo(newPosition))
                {
                    this.position = newPosition;
                }
            }
            case BACKWARD -> {
                Vector2d newPosition = this.position.subtract(this.direction.toUnitVector());

                if (validator.canMoveTo(newPosition))
                {
                    this.position = newPosition;
                }
            }
        }
    }

    public Vector2d getPosition() {
        return this.position;
    }

    public MapDirection getDirection() {
        return this.direction;
    }







  
}

File: labK/src/main/kotlin/agh/isc/oop/model/Animal.kt
package agh.ics.oop.model

import main.agh.ics.oop.model.MapDirection


class Animal(var position: Vector2d = Vector2d(2, 2), direction: MapDirection = MapDirection.NORTH){


    var direction: MapDirection = MapDirection.NORTH
        private set


    override fun toString(): String {
        return when(this.direction){
            MapDirection.NORTH -> "^"
            MapDirection.SOUTH -> "v"
            MapDirection.EAST -> ">"
            MapDirection.WEST -> "<"
        }
    }

    public fun isAt(position: Vector2d): Boolean {
        return this.position == position

    }

    public fun move(direction: MoveDirection, validator: WorldMap):Unit{
        when(direction)
        {
            MoveDirection.FORWARD -> {
                val newPosition = this.position + this.direction.toUnitVector()
                if(validator.canMoveTo(newPosition))
                {
                    this.position = newPosition
                }
            }
            MoveDirection.BACKWARD -> {
                val newPosition = this.position - this.direction.toUnitVector()
                if(validator.canMoveTo(newPosition))
                {
                    this.position = newPosition
                }
            }
            MoveDirection.RIGHT -> {
                this.direction = this.direction.next()
            }
            MoveDirection.LEFT -> {
                this.direction = this.direction.previous()
            }

        }

    }




}